pkg/controllers/provisioning/scheduling/topologynodefilter.go

```diff
@@ -27,102 +27,60 @@ import (
+	// if we only have a label selector, that's the only requirement that must match
+	if p.Spec.Affinity == nil || p.Spec.Affinity.NodeAffinity == nil || p.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil {
+		return TopologyNodeFilter{nodeSelectorRequirements}
 	}
-	return selector
-}
 
-// Matches returns true if the TopologyNodeFilter doesn't prohibit node from the participating in the topology
-func (t *TopologyNodeFilter) Matches(node *v1.Node) bool {
-	if t == nil {
-		return true
-	}
-	// if a node selector term is provided, it must match the node
-	for k, v := range t.nodeSelector {
-		if node.Labels[k] != v {
-			return false
-		}
-	}
-	if len(t.nodeSelectorTerms) == 0 {
-		return true
+	// otherwise, we need to match the combination of label selector and any term of the required node affinities since
+	// those terms are OR'd together
+	var filter TopologyNodeFilter
+	for _, term := range p.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
+		requirements := nodeSelectorRequirements.Add(v1alpha5.NewRequirements(term.MatchExpressions...).Requirements...)
+		filter = append(filter, requirements)
 	}
 
-	for _, term := range t.nodeSelectorTerms {
-		requirement := v1alpha5.NewRequirements(term.MatchExpressions...)
-		if t.matchesRequirements(requirement, node) {
-			return true
-		}
-	}
+	return filter
+}
 
-	// at this point we have node selector terms, but didn't match all of the requirements for any individual term
-	return false
+// Matches returns true if the TopologyNodeFilter doesn't prohibit node from the participating in the topology
+func (t TopologyNodeFilter) Matches(node *v1.Node) bool {
+	nodeLabels := v1alpha5.NewLabelRequirements(node.Labels)
+	return t.isCompatible(nodeLabels)
 }
 
 // MatchesRequirements returns true if the TopologyNodeFilter doesn't prohibit a node with the requirements from
 // participating in the topology. This method allows checking the requirements from a scheduling.Node to see if the
 // node we will soon create participates in this topology.
-func (t *TopologyNodeFilter) MatchesRequirements(nodeRequirements v1alpha5.Requirements) bool {
-	if t == nil {
-		return true
-	}
-	for k, v := range t.nodeSelector {
-		if !nodeRequirements.Get(k).Has(v) {
-			return false
-		}
-	}
-	if len(t.nodeSelectorTerms) == 0 {
-		return true
-	}
-
-	for _, term := range t.nodeSelectorTerms {
-		requirement := v1alpha5.NewRequirements(term.MatchExpressions...)
-		matchesAllReqs := true
-		for key := range requirement.Keys() {
-			if requirement.Get(key).Intersection(nodeRequirements.Get(key)).Len() == 0 {
-				matchesAllReqs = false
-				break
-			}
-		}
-		// these terms are OR'd together, so if we match one full set of requirements, the filter passes
-		if matchesAllReqs {
-			return true
-		}
-	}
-	return true
+func (t TopologyNodeFilter) MatchesRequirements(nodeRequirements v1alpha5.Requirements) bool {
```

Comment:
Why not just use the Compatible interface?

Reply:
I don't think it makes as much sense to ask if a filter is 'Compatible' with something compared to asking if a filter 'Matches' something.

Reply:
The intent of this comment is to reduce the number of concepts that a reader needs to be aware of. Naming of those concepts can happen later. It's not clear to me that we have multiple concepts we're dealing with.

Reply:
To me it's about readability of the function when used, the first is far more clear than the second:

```go
// nodes may or may not be considered for counting purposes for topology spread constraints depending on if they
// are selected by the pod's node selectors and required node affinities.  If these are unset, the node always counts.
if !tg.nodeFilter.Matches(node) {
        continue
}
```


```go
// nodes may or may not be considered for counting purposes for topology spread constraints depending on if they
// are selected by the pod's node selectors and required node affinities.  If these are unset, the node always counts.
if err := tg.nodeFilter.Compatible(node); err != nil  {
        continue
}
```